File: src/collab/rename.py
"""
Rename a collaborator's local directory and config entry.

Usage:
  corrkit for rename old-name new-name
"""

import argparse
import subprocess
import sys
from pathlib import Path

import resolve

from . import load_collaborators, save_collaborators


def _run(cmd: list[str], check: bool = True) -> subprocess.CompletedProcess:
    print(f"  $ {' '.join(cmd)}")
    result = subprocess.run(cmd, capture_output=True, text=True)
    if check and result.returncode != 0:
        print(f"  FAILED: {result.stderr.strip()}")
        sys.exit(1)
    return result


def _find_collab_dir(name: str) -> Path | None:
    """Find the collaborator directory (for/ or by/ inside correspondence)."""
    dd = resolve.data_dir()
    for prefix in [dd / "for", dd / "by"]:
        candidate = prefix / name.lower()
        if candidate.exists():
            return candidate
    return None


def main() -> None:
    parser = argparse.ArgumentParser(description="Rename a collaborator")
    parser.add_argument("old_name", help="Current collaborator name (TOML key)")
    parser.add_argument("new_name", help="New collaborator name")
    parser.add_argument(
        "--rename-repo",
        action="store_true",
        help="Also rename the GitHub repo to match the new name",
    )
    args = parser.parse_args()

    old_name: str = args.old_name
    new_name: str = args.new_name

    collabs = load_collaborators()

    if old_name not in collabs:
        print(f"Collaborator '{old_name}' not found in collaborators.toml")
        sys.exit(1)

    if new_name in collabs:
        print(f"Collaborator '{new_name}' already exists in collaborators.toml")
        sys.exit(1)

    collab = collabs[old_name]

    # 1. Move directory via git mv (if it exists)
    old_dir = _find_collab_dir(old_name)
    if old_dir is not None:
        new_dir = old_dir.parent / new_name.lower()
        print(f"Moving {old_dir} → {new_dir}")
        _run(["git", "mv", str(old_dir), str(new_dir)])
    else:
        print(f"Directory for '{old_name}' not found on disk — skipping git mv")

    # 2. Optionally rename the GitHub repo
    if args.rename_repo and collab.repo:
        try:
            from accounts import load_owner

            owner_gh = load_owner().github_user
        except (SystemExit, ImportError):
            owner_gh = ""
        new_repo_name = f"to-{new_name.lower()}"
        print(f"Renaming GitHub repo {collab.repo} → {new_repo_name}")
        _run(["gh", "repo", "rename", new_repo_name, "-R", collab.repo, "--yes"])
        # Update repo field to reflect the new name
        new_repo = f"{owner_gh}/{new_repo_name}" if owner_gh else new_repo_name
    else:
        new_repo = collab.repo

    # 3. Update collaborators.toml
    import msgspec.structs

    updated = msgspec.structs.replace(collab, github_user=new_name, repo=new_repo)
    del collabs[old_name]
    collabs[new_name] = updated
    save_collaborators(collabs)
    print(f"Renamed '{old_name}' → '{new_name}' in collaborators.toml")

    print(f"Done. Collaborator '{old_name}' renamed to '{new_name}'.")


if __name__ == "__main__":
    main()

File: src/mailbox/rename.rs
//! Rename a mailbox directory and config entry.

use anyhow::Result;

use crate::resolve;
use crate::util::run_cmd_checked;

pub fn run(old_name: &str, new_name: &str, rename_repo: bool) -> Result<()> {
    let old_dir = resolve::mailbox_dir(old_name);
    let new_dir = resolve::mailbox_dir(new_name);

    if new_dir.exists() {
        anyhow::bail!("Mailbox '{}' already exists at {}", new_name, new_dir.display());
    }

    // 1. Move directory via git mv (if it exists)
    if old_dir.exists() {
        println!(
            "Moving {} \u{2192} {}",
            old_dir.display(),
            new_dir.display()
        );

        // Ensure parent directory exists
        if let Some(parent) = new_dir.parent() {
            std::fs::create_dir_all(parent)?;
        }

        run_cmd_checked(&[
            "git",
            "mv",
            &old_dir.to_string_lossy(),
            &new_dir.to_string_lossy(),
        ])?;
    } else {
        println!(
            "Directory for '{}' not found on disk \u{2014} skipping git mv",
            old_name
        );
    }

    // 2. Optionally rename the GitHub repo
    if rename_repo {
        let owner_gh = crate::accounts::load_owner(None)
            .map(|o| o.github_user)
            .unwrap_or_default();
        if !owner_gh.is_empty() {
            let old_repo = format!("{}/to-{}", owner_gh, old_name.to_lowercase());
            let new_repo_name = format!("to-{}", new_name.to_lowercase());
            println!(
                "Renaming GitHub repo {} \u{2192} {}",
                old_repo, new_repo_name
            );
            run_cmd_checked(&[
                "gh",
                "repo",
                "rename",
                &new_repo_name,
                "-R",
                &old_repo,
                "--yes",
            ])?;
        }
    }

    // 3. Update .corky.toml
    update_config(old_name, new_name)?;

    println!(
        "Done. Mailbox '{}' renamed to '{}'.",
        old_name, new_name
    );
    Ok(())
}

/// Rename mailbox in .corky.toml: update [mailboxes] key and [routing] paths.
fn update_config(old_name: &str, new_name: &str) -> Result<()> {
    let config_path = resolve::corky_toml();
    if !config_path.exists() {
        return Ok(());
    }

    let content = std::fs::read_to_string(&config_path)?;
    let mut doc = content.parse::<toml_edit::DocumentMut>()?;

    // Rename [mailboxes.{old}] → [mailboxes.{new}]
    if let Some(mailboxes) = doc.get_mut("mailboxes") {
        if let Some(table) = mailboxes.as_table_mut() {
            if let Some(entry) = table.remove(old_name) {
                table.insert(new_name, entry);
            }
        }
    }

    // Update routing paths
    let old_path = format!("mailboxes/{}", old_name);
    let new_path = format!("mailboxes/{}", new_name);
    if let Some(routing) = doc.get_mut("routing") {
        if let Some(table) = routing.as_table_mut() {
            for (_, item) in table.iter_mut() {
                if let Some(arr) = item.as_array_mut() {
                    for i in 0..arr.len() {
                        if arr.get(i).and_then(|v| v.as_str()) == Some(&old_path) {
                            arr.replace(i, &new_path);
                        }
                    }
                }
            }
        }
    }

    std::fs::write(&config_path, doc.to_string())?;
    println!(
        "Renamed '{}' \u{2192} '{}' in .corky.toml",
        old_name, new_name
    );

    Ok(())
}
